crates/matrix-sdk/src/client/builder.rs

```diff
@@ -371,63 +402,88 @@ impl ClientBuilder {
+                            Some(RequestConfig::short_retry()),
+                            homeserver_url.to_string(),
+                            None,
+                            &[MatrixVersion::V1_0],
+                            Default::default(),
+                        )
+                        .await
+                        .is_ok()
+                    {
+                        result = Ok((homeserver_url.to_string(), None));
+                    } else {
+                        // If it isn't, we'll throw the discovery error instead.
+                    }
                 }
 
-                url
+                result?
             }
-            HomeserverConfig::ServerName { server: server_name, protocol } => {
-                debug!("Trying to discover the homeserver");
-
-                let homeserver = match protocol {
-                    UrlScheme::Http => format!("http://{server_name}"),
-                    UrlScheme::Https => format!("https://{server_name}"),
-                };
-
-                let well_known = http_client
-                    .send(
-                        discover_homeserver::Request::new(),
-                        Some(RequestConfig::short_retry()),
-                        homeserver,
-                        None,
-                        &[MatrixVersion::V1_0],
-                        Default::default(),
-                    )
-                    .await
-                    .map_err(|e| match e {
-                        HttpError::Api(err) => ClientBuildError::AutoDiscovery(err),
-                        err => ClientBuildError::Http(err),
-                    })?;
-
-                #[cfg(feature = "experimental-oidc")]
-                {
-                    authentication_server_info = well_known.authentication;
-                    allow_insecure_oidc = matches!(protocol, UrlScheme::Http);
-                }
-
-                #[cfg(feature = "experimental-sliding-sync")]
-                if let Some(proxy) = well_known.sliding_sync_proxy.map(|p| p.url) {
-                    sliding_sync_proxy = Url::parse(&proxy).ok();
-                }
+        };
 
-                debug!(
-                    homeserver_url = well_known.homeserver.base_url,
-                    "Discovered the homeserver"
-                );
+        #[allow(unused_variables)]
+        if let Some(well_known) = well_known {
+            #[cfg(feature = "experimental-oidc")]
+            {
+                authentication_server_info = well_known.authentication;
+            }
 
-                well_known.homeserver.base_url
+            #[cfg(feature = "experimental-sliding-sync")]
+            if let Some(proxy) = well_known.sliding_sync_proxy.map(|p| p.url) {
+                sliding_sync_proxy = Url::parse(&proxy).ok();
             }
-        };
+        } else {
+            #[cfg(feature = "experimental-sliding-sync")]
+            {
+                sliding_sync_proxy =
```

Comment:
The comment around the method `ClientBuilder::sliding_sync_proxy()` needs an update too! And since we're going to modify it: can we also clarify in the comment there that the value set with this function may be override with a value automatically discovered through `.well-known`?

Reply:
Updated, but to mention the change I mentioned below.